File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.cc
#include <iostream>
#define MAX(x, y) ((x) > (y) ? (x) : (y))

using namespace std;

int main(void) {
	int n, ans{ 0 };

	cin >> n;
	int* arr = new int[n + 1]();
	int* dp = new int[n + 1]();

	for (int i = 1; i <= n; i++) {
		cin >> arr[i];

		int max = 0;
		for (int j = 0; j < i; j++) {
			if (arr[j] < arr[i] && max < dp[j]) {
				max = dp[j];
			}
		}
		dp[i] = max + 1;
		ans = MAX(ans, dp[i]);
	}

	cout << ans << endl;

	return 0;
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt(), ans = 0;
        int[] arr = new int[n + 1];
        int[] dp = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            arr[i] = sc.nextInt();

            int max = 0;
            for (int j = 0; j < i; j++) {
                if(arr[j] < arr[i] && max < dp[j]){
                    max = dp[j];
                }
            }
            dp[i] = max + 1;
            ans = Math.max(ans, dp[i]);
        }

        System.out.println(ans);
    }
}
